xdsl/dialects/builtin.py

```diff
@@ -353,8 +353,31 @@ def print_parameter(self, printer: Printer) -> None:
                 raise ValueError(f"Invalid signedness {data}")
 
 
+class FixedBitwidthType(TypeAttribute, ABC):
+    """
+    A type attribute with a defined bitwidth
+    """
+
+    name = "abstract.bitwidth_type"
+
+    @property
+    @abstractmethod
+    def bitwidth(self) -> int:
+        """
+        Contiguous memory footprint in bits
+        """
+        raise NotImplementedError()
+
+    @property
+    def size(self) -> int:
+        """
+        Contiguous memory footprint in bytes, defaults to `ceil(bitwidth / 8)`
+        """
+        return ceil(self.bitwidth / 8)
```

Comment:
This returns a float. I would do the weird int modulo thing tbh, since the bitwidth is guaranteed to be positive, and division is expensive
```suggestion
        bitwidth = self.bitwidth
        return bitwidth >> 3 + bool(bitwidth % 8)
```